### ISSUE TITLE
`duplicated_attributes` false positive on lint reasons (and other attributes too)

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#13355]
### Summary

When two lint reasons are the same, `duplicated_attributes` lints and suggests that one of them should be removed. This would change the meaning.

Besides the specific case of `allow(..., reason = ...)`, I also observe that [the implementation](https://github.com/rust-lang/rust-clippy/blob/9e9042a11074c83a2c40b4e8ce29f9e2bb610d34/clippy_lints/src/attrs/duplicated_attributes.rs#L57-L61) seems to be making the assumption that, unless otherwise specified, all attributes’ syntax can be interpreted as a tree of shared prefixes like `use foo::{bar, baz}` is. That's not true; it's up to the attribute’s definition how its contents are interpreted. If that false assumption were removed, then #13238 would also be fixed, and #12619 (`rustc_on_unimplemented`) wouldn't need a special case.

### Lint Name

`duplicated_attributes`

### Reproducer

I tried this code:

```rust
#![allow(missing_docs, reason = "library for internal use only")]
#![allow(
    exported_private_dependencies,
    reason = "library for internal use only"
)]
```

I saw this happen:

```
warning: duplicated attribute
 --> src/lib.rs:4:5
  |
4 |     reason = "library for internal use only"
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: first defined here
 --> src/lib.rs:1:24
  |
1 | #![allow(missing_docs, reason = "library for internal use only")]
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: remove this attribute
 --> src/lib.rs:4:5
  |
4 |     reason = "library for internal use only"
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#duplicated_attributes
  = note: `#[warn(clippy::duplicated_attributes)]` on by default
```

Expected: No lint

### Version

```text
rustc 1.83.0-nightly (4ac7bcbaa 2024-09-04)
binary: rustc
commit-hash: 4ac7bcbaad8d6fd7a51bdf1b696cbc3ba4c796cf
commit-date: 2024-09-04
host: x86_64-apple-darwin
release: 1.83.0-nightly
LLVM version: 19.1.0
```

**Comment by samueltardieu** [comment]
This has been fixed.

**Comment by MareStare** [comment]
> https://github.com/rust-lang/rust-clippy/issues/12619 (rustc_on_unimplemented) wouldn't need a special case.

True! `bon` has an attribute with the semantics very similar to `rustc_on_unimplemented` causing a false-positive lint:

```rust
#[derive(bon::Builder)]
#[builder(on(String, into), on(PathBuf, into))]
pub struct Example {}
```

Here, double `into` is fine according to [`#[builder(on())]`](https://bon-rs.com/reference/builder/top-level/on) semantics.

Warning:

```
warning: duplicated attribute
 --> crates/sandbox/src/main.rs:2:41
  |
2 | #[builder(on(String, into), on(PathBuf, into))]
  |                                         ^^^^
  |
note: first defined here
 --> crates/sandbox/src/main.rs:2:22
  |
2 | #[builder(on(String, into), on(PathBuf, into))]
  |                      ^^^^
help: remove this attribute
 --> crates/sandbox/src/main.rs:2:41
  |
2 | #[builder(on(String, into), on(PathBuf, into))]
  |                                         ^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#duplicated_attributes
  = note: `#[warn(clippy::duplicated_attributes)]` on by default
```

**Comment by Veetaha** [comment]
> https://github.com/rust-lang/rust-clippy/issues/12619 (rustc_on_unimplemented) wouldn't need a special case.

True! `bon` has an attribute with the semantics very similar to `rustc_on_unimplemented` causing a false-positive lint:

```rust
#[derive(bon::Builder)]
#[builder(on(String, into), on(PathBuf, into))]
pub struct Example {}
```

Here, double `into` is fine according to [`#[builder(on())]`](https://bon-rs.com/reference/builder/top-level/on) semantics.

Warning:

```
warning: duplicated attribute
 --> crates/sandbox/src/main.rs:2:41
  |
2 | #[builder(on(String, into), on(PathBuf, into))]
  |                                         ^^^^
  |
note: first defined here
 --> crates/sandbox/src/main.rs:2:22
  |
2 | #[builder(on(String, into), on(PathBuf, into))]
  |                      ^^^^
help: remove this attribute
 --> crates/sandbox/src/main.rs:2:41
  |
2 | #[builder(on(String, into), on(PathBuf, into))]
  |                                         ^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#duplicated_attributes
  = note: `#[warn(clippy::duplicated_attributes)]` on by default
```

@samueltardieu I tried the latest nightly and I still see a false-positive for `bon`'s attributes

```
cargo 1.88.0-nightly (864f74d4e 2025-04-11)
```
